engine/graphics/src/opengl/graphics_opengl.cpp

```diff
@@ -1751,54 +1815,153 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
         CHECK_GL_ERROR
     }
 
-    static uint32_t CreateShader(GLenum type, const void* program, uint32_t program_size)
+    static GLuint DoCreateShader(GLenum type, const void* program, uint32_t program_size)
     {
-        GLuint s = glCreateShader(type);
+        GLuint shader_id = glCreateShader(type);
         CHECK_GL_ERROR;
         GLint size = program_size;
-        glShaderSource(s, 1, (const GLchar**) &program, &size);
+        glShaderSource(shader_id, 1, (const GLchar**) &program, &size);
         CHECK_GL_ERROR;
-        glCompileShader(s);
+        glCompileShader(shader_id);
         CHECK_GL_ERROR;
 
         GLint status;
-        glGetShaderiv(s, GL_COMPILE_STATUS, &status);
+        glGetShaderiv(shader_id, GL_COMPILE_STATUS, &status);
         if (status == 0)
         {
 #ifndef NDEBUG
             GLint logLength;
-            glGetShaderiv(s, GL_INFO_LOG_LENGTH, &logLength);
+            glGetShaderiv(shader_id, GL_INFO_LOG_LENGTH, &logLength);
             if (logLength > 0)
             {
                 GLchar *log = (GLchar *)malloc(logLength);
-                glGetShaderInfoLog(s, logLength, &logLength, log);
+                glGetShaderInfoLog(shader_id, logLength, &logLength, log);
                 dmLogError("%s\n", log);
                 free(log);
             }
 #endif
-            glDeleteShader(s);
+            glDeleteShader(shader_id);
             return 0;
         }
 
-        return s;
+        return shader_id;
+    }
+
+    static Shader* CreateShader(GLenum type, ShaderDesc::Shader* ddf)
+    {
+        // Note: The DDF will be removed after the shader has been created
+        //       so we need to store a copy of the texture array source
+        //       since the programs that use this variant will patch in the
+        //       maximum number of pages supported before creating the 
+        //       shader program.
+        if (ddf->m_VariantTextureArray)
```

Comment:
Can we move this somewhere outside the graphics library, like the material or via the build system?
* material - we need to know when to do this source transformation and currently the material can't get that information from the built resource since it's just a shader handle. We could store it on the shader itself and defer with shader creation until the program gets created or something like that..
* build system - the material just references the fragment resource, so in this case we would need to output a second built resource for every material that has this variant..